app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java

```diff
@@ -0,0 +1,119 @@
+                UserError.Log.d(TAG, "Treatment count request for: " + count + " entries");
+            } catch (NumberFormatException e) {
+                UserError.Log.wtf(TAG, "Invalid treatment count request for: " + count + " entries");
+            }
+        }
+
+        // Store a cache of the last Treatments.latest() query for the duration in which there is no
+        // new latest treatment. Since obtaining the latest treatment is fast, but a larger number of
+        // treatments is significantly slower, this optimizes the most often use case.
+        Treatments latestTreatment = null;
+        List<Treatments> latestList = Treatments.latest(1);
+        if (latestList != null && !latestList.isEmpty()) {
+            latestTreatment = latestList.iterator().next();
+        }
+
+        // Check whether there are any cached treatments, and if so whether the latest cached
+        // treatment matches the UUID of the most recent treatment. Use the cache if we can,
+        // otherwise fill cachedTreatments with new treatments from the database.
+        List<Treatments> treatments;
+        if (cachedTreatments != null && cachedTreatments.size() > 0 && count <= cachedTreatments.size() &&
+            latestTreatment != null && latestTreatment.uuid.equals(cachedTreatments.iterator().next().uuid))
+        {
+            if (count == cachedTreatments.size()) {
+                UserError.Log.d(TAG, "Using cached treatments");
+                treatments = cachedTreatments;
+            } else {
+                UserError.Log.d(TAG, "Copying " + count + " of " + cachedTreatments.size() + " cached treatments");
+                treatments = new ArrayList<>();
+                Iterator<Treatments> it = cachedTreatments.iterator();
+                for (int i=0; i<count; i++) {
+                    treatments.add(it.next());
+                }
+            }
+        } else {
+            UserError.Log.d(TAG, "Fetching latest " + count + " entries from Treatments");
+            treatments = Treatments.latest(count);
+            cachedTreatments = treatments;
+        }
+
+        final JSONArray reply = new JSONArray();
+
+        // populate json structures
+        try {
+            for (Treatments treatment : treatments) {
+                final JSONObject item = new JSONObject();
+
+                item.put("_id", treatment.uuid);
+                item.put("created_at", treatment.timestamp);
+                item.put("eventType", treatment.eventType);
+                item.put("enteredBy", treatment.enteredBy);
+                item.put("notes", treatment.notes);
+                item.put("carbs", treatment.carbs);
+                item.put("insulin", treatment.insulin);
+
+                reply.put(item);
+            }
+
+            Log.d(TAG, "Treatment output: " + reply.toString());
+        } catch (JSONException e) {
+            UserError.Log.wtf(TAG, "Got json exception: " + e);
```

Comment:
Again, not a terrible failure which breaks xDrip, just the webservice response.

Reply:
Changed to w